### ISSUE TITLE
CORS support

### ISSUE THREAD

**Issue by ikkez** [bcosca/fatfree#731]
Hi there.
Would be nice to have some support for Cross-Origin Resource Sharing ([CORS@wikipedia](http://en.wikipedia.org/wiki/Cross-origin_resource_sharing)) in the core that adds some juggling with required headers for REST-API apps, especially for request over AJAX, where those cross-origin requests are blocked by default (like jQuery.ajax from `domain-a.com` requests to `domain-b.com` (F3-REST)).
To get it working, you need to set the `Access-Control-Allow-Origin: *` header for outgoing responses.

Additionally performing a `PUT` request to the server will add another preflight `OPTIONS` request in conjunction, that will send a `Access-Control-Request-Method: PUT` header and expect a response header containing something like `Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS`. ([SO](http://stackoverflow.com/questions/17702220/cross-domain-jquery-ajax-request-fails-for-put-method-put-is-not-allowed-by-a)) 
The current code already supports this for regular requests and sends this as `Allow: ...`.
The `Access-Control-Allow-Origin: *` header is no problem to add in own controllers, but for the PUT/OPTIONS thing it would fit perfectly in here: https://github.com/bcosca/fatfree-core/blob/master/base.php#L1449
So basically it's just a copy of the line before

```php
header('Access-Control-Allow-Methods: '.implode(',',$allowed));
```

But maybe some checks and hive configs for this would be nice. Currently this works fine here, but i found this flowchart for implementing CORS (on [enable-cors.org](http://enable-cors.org/server.html)) which does a little bit more, which might be considered:

![flowchart for implementing CORS](http://www.html5rocks.com/static/images/cors_server_flowchart.png)

thanks.

**Comment by exophunk** [comment]
Is this issue going to be solved in the future?

I'm building a RESTful API now which is called from a React/Flux based JS Client. The Fat-Free API get called via AJAX requests and coming from a different origin, so CORS need to be enabled.

This usually works fine by just adding a `header('Access-Control-Allow-Origin: *');` in the controller class, BUT: I tried to add a **custom HTTP header** to my ajax request, called "token", because I want to pass a JWT to the API on each request. In this case, the browser preflights the request and first sends OPTIONS to the server. There is no practical way to bypass this.

Fat-Free in this case doesn't provide any way to include `header('Access-Control-Allow-Origin: *');` for the OPTIONS method and therefore any preflighted AJAX request fails due to CORS restrictions.
`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource [...]`

As mentioned above by @ikkez, this can be bypassed by allowing CORS right before this line: https://github.com/bcosca/fatfree-core/blob/master/base.php#L1510

In my case I added the following lines of code to allow CORS and also allow my custom header:
```
if ($this->hive['VERB']=='OPTIONS')
				header('Access-Control-Allow-Origin: *');
				header('Access-Control-Allow-Headers: token');
```

Obviously, this cannot be added to the framework directly, but there need to be an option to specify allowed HTTP headers and Origins. It would be nice to just configure this via the F3 config.

Thanks!
